From b96ff93fdfd018c1fd01ae44ed3a42587b706496 Mon Sep 17 00:00:00 2001
From: Yaroslav Zviezda <acroreiser@gmail.com>
Date: Wed, 29 Dec 2021 11:41:03 +0300
Subject: [PATCH] sepolicy: allow su domain in user builds

Do not use for anything except boot time one shot scripts.

Change-Id: I3ac61e9e3e9741d512ece95c86a8e7ee1617a926
---
 build/soong/policy.go                |  4 ++--
 prebuilts/api/202404/private/init.te | 10 +++++-----
 prebuilts/api/202404/private/su.te   |  2 --
 prebuilts/api/202504/private/init.te |  3 +--
 prebuilts/api/202504/private/su.te   |  3 +--
 prebuilts/api/30.0/private/init.te   | 10 +++++-----
 prebuilts/api/30.0/private/su.te     |  2 --
 prebuilts/api/31.0/private/init.te   | 10 +++++-----
 prebuilts/api/31.0/private/su.te     |  3 ---
 prebuilts/api/32.0/private/init.te   | 10 +++++-----
 prebuilts/api/32.0/private/su.te     |  3 ---
 prebuilts/api/33.0/private/init.te   | 10 +++++-----
 prebuilts/api/33.0/private/su.te     |  3 ---
 prebuilts/api/34.0/private/init.te   | 10 +++++-----
 prebuilts/api/34.0/private/su.te     |  2 --
 private/init.te                      | 10 +++++-----
 private/su.te                        |  2 --
 17 files changed, 39 insertions(+), 58 deletions(-)

diff --git a/build/soong/policy.go b/build/soong/policy.go
index 23ff255..c34eb81 100644
--- a/build/soong/policy.go
+++ b/build/soong/policy.go
@@ -558,7 +558,7 @@ func (c *policyBinary) GenerateAndroidBuildActions(ctx android.ModuleContext) {
 		rule.Temporary(permissiveDomains)
 
 		msg := `==========\n` +
-			`ERROR: permissive domains not allowed in user builds\n` +
+			`WARNING: permissive domains not allowed in user builds\n` +
 			`List of invalid domains:`
 
 		rule.Command().Text("if test").
@@ -568,7 +568,7 @@ func (c *policyBinary) GenerateAndroidBuildActions(ctx android.ModuleContext) {
 			Text(`"` + msg + `"`).
 			Text("&& cat ").
 			Input(permissiveDomains).
-			Text("; exit 1; fi")
+			Text("; fi")
 	}
 
 	out := pathForModuleOut(ctx, c.stem())
diff --git a/prebuilts/api/202404/private/init.te b/prebuilts/api/202404/private/init.te
index 9d3a2c3..f34d166 100644
--- a/prebuilts/api/202404/private/init.te
+++ b/prebuilts/api/202404/private/init.te
@@ -28,13 +28,13 @@ domain_trans(init, { rootfs toolbox_exec }, modprobe)
 userdebug_or_eng(`
   # case where logpersistd is actually logcat -f in logd context (nee: logcatd)
   domain_auto_trans(init, logcat_exec, logpersist)
-
-  # allow init to execute services marked with seclabel u:r:su:s0 in userdebug/eng
-  allow init su:process transition;
-  dontaudit init su:process noatsecure;
-  allow init su:process { siginh rlimitinh };
 ')
 
+# allow init to execute services marked with seclabel u:r:su:s0 in userdebug/eng
+allow init su:process transition;
+dontaudit init su:process noatsecure;
+allow init su:process { siginh rlimitinh };
+
 # Allow init to figure out name of dm-device from it's /dev/block/dm-XX path.
 # This is useful in case of remounting ext4 userdata into checkpointing mode,
 # since it potentially requires tearing down dm-devices (e.g. dm-bow, dm-crypto)
diff --git a/prebuilts/api/202404/private/su.te b/prebuilts/api/202404/private/su.te
index 2e0d10a..5b5cd3d 100644
--- a/prebuilts/api/202404/private/su.te
+++ b/prebuilts/api/202404/private/su.te
@@ -1,4 +1,3 @@
-userdebug_or_eng(`
   typeattribute su coredomain;
 
   domain_auto_trans(shell, su_exec, su)
@@ -33,4 +32,3 @@ userdebug_or_eng(`
 
   # Allow root to set MTE permissive mode.
   set_prop(su, permissive_mte_prop);
-')
diff --git a/prebuilts/api/202504/private/init.te b/prebuilts/api/202504/private/init.te
index 6f0ee80..ee900f8 100644
--- a/prebuilts/api/202504/private/init.te
+++ b/prebuilts/api/202504/private/init.te
@@ -28,12 +28,11 @@ domain_trans(init, { rootfs toolbox_exec }, modprobe)
 userdebug_or_eng(`
   # case where logpersistd is actually logcat -f in logd context (nee: logcatd)
   domain_auto_trans(init, logcat_exec, logpersist)
-
+')
   # allow init to execute services marked with seclabel u:r:su:s0 in userdebug/eng
   allow init su:process transition;
   dontaudit init su:process noatsecure;
   allow init su:process { siginh rlimitinh };
-')
 
 # Allow init to figure out name of dm-device from it's /dev/block/dm-XX path.
 # This is useful in case of remounting ext4 userdata into checkpointing mode,
diff --git a/prebuilts/api/202504/private/su.te b/prebuilts/api/202504/private/su.te
index 247fd0b..a02100c 100644
--- a/prebuilts/api/202504/private/su.te
+++ b/prebuilts/api/202504/private/su.te
@@ -1,4 +1,3 @@
-userdebug_or_eng(`
   typeattribute su coredomain;
 
   domain_auto_trans(shell, su_exec, su)
@@ -133,4 +132,4 @@ userdebug_or_eng(`
   typeattribute su hal_wifi_client;
   typeattribute su hal_wifi_hostapd_client;
   typeattribute su hal_wifi_supplicant_client;
-')
+
diff --git a/prebuilts/api/30.0/private/init.te b/prebuilts/api/30.0/private/init.te
index b0e7f80..e40b159 100644
--- a/prebuilts/api/30.0/private/init.te
+++ b/prebuilts/api/30.0/private/init.te
@@ -24,13 +24,13 @@ domain_trans(init, { rootfs toolbox_exec }, modprobe)
 userdebug_or_eng(`
   # case where logpersistd is actually logcat -f in logd context (nee: logcatd)
   domain_auto_trans(init, logcat_exec, logpersist)
-
-  # allow init to execute services marked with seclabel u:r:su:s0 in userdebug/eng
-  allow init su:process transition;
-  dontaudit init su:process noatsecure;
-  allow init su:process { siginh rlimitinh };
 ')
 
+# allow init to execute services marked with seclabel u:r:su:s0 in userdebug/eng
+allow init su:process transition;
+dontaudit init su:process noatsecure;
+allow init su:process { siginh rlimitinh };
+
 # Allow init to figure out name of dm-device from it's /dev/block/dm-XX path.
 # This is useful in case of remounting ext4 userdata into checkpointing mode,
 # since it potentially requires tearing down dm-devices (e.g. dm-bow, dm-crypto)
diff --git a/prebuilts/api/30.0/private/su.te b/prebuilts/api/30.0/private/su.te
index 16e47bb..8ed6ee8 100644
--- a/prebuilts/api/30.0/private/su.te
+++ b/prebuilts/api/30.0/private/su.te
@@ -1,4 +1,3 @@
-userdebug_or_eng(`
   typeattribute su coredomain;
 
   domain_auto_trans(shell, su_exec, su)
@@ -20,4 +19,3 @@ userdebug_or_eng(`
   permissive su;
 
   app_domain(su)
-')
diff --git a/prebuilts/api/31.0/private/init.te b/prebuilts/api/31.0/private/init.te
index f569e0c..7eac8d7 100644
--- a/prebuilts/api/31.0/private/init.te
+++ b/prebuilts/api/31.0/private/init.te
@@ -25,13 +25,13 @@ domain_trans(init, { rootfs toolbox_exec }, modprobe)
 userdebug_or_eng(`
   # case where logpersistd is actually logcat -f in logd context (nee: logcatd)
   domain_auto_trans(init, logcat_exec, logpersist)
-
-  # allow init to execute services marked with seclabel u:r:su:s0 in userdebug/eng
-  allow init su:process transition;
-  dontaudit init su:process noatsecure;
-  allow init su:process { siginh rlimitinh };
 ')
 
+# allow init to execute services marked with seclabel u:r:su:s0 in userdebug/eng
+allow init su:process transition;
+dontaudit init su:process noatsecure;
+allow init su:process { siginh rlimitinh };
+
 # Allow init to figure out name of dm-device from it's /dev/block/dm-XX path.
 # This is useful in case of remounting ext4 userdata into checkpointing mode,
 # since it potentially requires tearing down dm-devices (e.g. dm-bow, dm-crypto)
diff --git a/prebuilts/api/31.0/private/su.te b/prebuilts/api/31.0/private/su.te
index 587f449..73e8944 100644
--- a/prebuilts/api/31.0/private/su.te
+++ b/prebuilts/api/31.0/private/su.te
@@ -1,4 +1,3 @@
-userdebug_or_eng(`
   typeattribute su coredomain;
 
   domain_auto_trans(shell, su_exec, su)
@@ -26,5 +25,3 @@ userdebug_or_eng(`
 
   # Do not audit accesses to keystore2 namespace for the su domain.
   dontaudit su keystore2_key_type:{ keystore2 keystore2_key } *;
-
-')
diff --git a/prebuilts/api/32.0/private/init.te b/prebuilts/api/32.0/private/init.te
index 200780d..2361044 100644
--- a/prebuilts/api/32.0/private/init.te
+++ b/prebuilts/api/32.0/private/init.te
@@ -25,13 +25,13 @@ domain_trans(init, { rootfs toolbox_exec }, modprobe)
 userdebug_or_eng(`
   # case where logpersistd is actually logcat -f in logd context (nee: logcatd)
   domain_auto_trans(init, logcat_exec, logpersist)
-
-  # allow init to execute services marked with seclabel u:r:su:s0 in userdebug/eng
-  allow init su:process transition;
-  dontaudit init su:process noatsecure;
-  allow init su:process { siginh rlimitinh };
 ')
 
+# allow init to execute services marked with seclabel u:r:su:s0 in userdebug/eng
+allow init su:process transition;
+dontaudit init su:process noatsecure;
+allow init su:process { siginh rlimitinh };
+
 # Allow init to figure out name of dm-device from it's /dev/block/dm-XX path.
 # This is useful in case of remounting ext4 userdata into checkpointing mode,
 # since it potentially requires tearing down dm-devices (e.g. dm-bow, dm-crypto)
diff --git a/prebuilts/api/32.0/private/su.te b/prebuilts/api/32.0/private/su.te
index 587f449..73e8944 100644
--- a/prebuilts/api/32.0/private/su.te
+++ b/prebuilts/api/32.0/private/su.te
@@ -1,4 +1,3 @@
-userdebug_or_eng(`
   typeattribute su coredomain;
 
   domain_auto_trans(shell, su_exec, su)
@@ -26,5 +25,3 @@ userdebug_or_eng(`
 
   # Do not audit accesses to keystore2 namespace for the su domain.
   dontaudit su keystore2_key_type:{ keystore2 keystore2_key } *;
-
-')
diff --git a/prebuilts/api/33.0/private/init.te b/prebuilts/api/33.0/private/init.te
index 17e25f8..c738473 100644
--- a/prebuilts/api/33.0/private/init.te
+++ b/prebuilts/api/33.0/private/init.te
@@ -26,13 +26,13 @@ domain_trans(init, { rootfs toolbox_exec }, modprobe)
 userdebug_or_eng(`
   # case where logpersistd is actually logcat -f in logd context (nee: logcatd)
   domain_auto_trans(init, logcat_exec, logpersist)
-
-  # allow init to execute services marked with seclabel u:r:su:s0 in userdebug/eng
-  allow init su:process transition;
-  dontaudit init su:process noatsecure;
-  allow init su:process { siginh rlimitinh };
 ')
 
+# allow init to execute services marked with seclabel u:r:su:s0 in userdebug/eng
+allow init su:process transition;
+dontaudit init su:process noatsecure;
+allow init su:process { siginh rlimitinh };
+
 # Allow init to figure out name of dm-device from it's /dev/block/dm-XX path.
 # This is useful in case of remounting ext4 userdata into checkpointing mode,
 # since it potentially requires tearing down dm-devices (e.g. dm-bow, dm-crypto)
diff --git a/prebuilts/api/33.0/private/su.te b/prebuilts/api/33.0/private/su.te
index 587f449..73e8944 100644
--- a/prebuilts/api/33.0/private/su.te
+++ b/prebuilts/api/33.0/private/su.te
@@ -1,4 +1,3 @@
-userdebug_or_eng(`
   typeattribute su coredomain;
 
   domain_auto_trans(shell, su_exec, su)
@@ -26,5 +25,3 @@ userdebug_or_eng(`
 
   # Do not audit accesses to keystore2 namespace for the su domain.
   dontaudit su keystore2_key_type:{ keystore2 keystore2_key } *;
-
-')
diff --git a/prebuilts/api/34.0/private/init.te b/prebuilts/api/34.0/private/init.te
index 9d3a2c3..f34d166 100644
--- a/prebuilts/api/34.0/private/init.te
+++ b/prebuilts/api/34.0/private/init.te
@@ -28,13 +28,13 @@ domain_trans(init, { rootfs toolbox_exec }, modprobe)
 userdebug_or_eng(`
   # case where logpersistd is actually logcat -f in logd context (nee: logcatd)
   domain_auto_trans(init, logcat_exec, logpersist)
-
-  # allow init to execute services marked with seclabel u:r:su:s0 in userdebug/eng
-  allow init su:process transition;
-  dontaudit init su:process noatsecure;
-  allow init su:process { siginh rlimitinh };
 ')
 
+# allow init to execute services marked with seclabel u:r:su:s0 in userdebug/eng
+allow init su:process transition;
+dontaudit init su:process noatsecure;
+allow init su:process { siginh rlimitinh };
+
 # Allow init to figure out name of dm-device from it's /dev/block/dm-XX path.
 # This is useful in case of remounting ext4 userdata into checkpointing mode,
 # since it potentially requires tearing down dm-devices (e.g. dm-bow, dm-crypto)
diff --git a/prebuilts/api/34.0/private/su.te b/prebuilts/api/34.0/private/su.te
index bac0ae0..fcb9822 100644
--- a/prebuilts/api/34.0/private/su.te
+++ b/prebuilts/api/34.0/private/su.te
@@ -1,6 +1,5 @@
 typeattribute su coredomain;
 
-userdebug_or_eng(`
   domain_auto_trans(shell, su_exec, su)
   # Allow dumpstate to call su on userdebug / eng builds to collect
   # additional information.
@@ -29,7 +28,6 @@ userdebug_or_eng(`
 
   # Allow root to set MTE permissive mode.
   set_prop(su, permissive_mte_prop);
-')
 
 # su is also permissive to permit setenforce.
 permissive su;
diff --git a/private/init.te b/private/init.te
index 31c4103..eb96c66 100644
--- a/private/init.te
+++ b/private/init.te
@@ -28,13 +28,13 @@ domain_trans(init, { rootfs toolbox_exec }, modprobe)
 userdebug_or_eng(`
   # case where logpersistd is actually logcat -f in logd context (nee: logcatd)
   domain_auto_trans(init, logcat_exec, logpersist)
-
-  # allow init to execute services marked with seclabel u:r:su:s0 in userdebug/eng
-  allow init su:process transition;
-  dontaudit init su:process noatsecure;
-  allow init su:process { siginh rlimitinh };
 ')
 
+# allow init to execute services marked with seclabel u:r:su:s0 in userdebug/eng
+allow init su:process transition;
+dontaudit init su:process noatsecure;
+allow init su:process { siginh rlimitinh };
+
 # Allow init to figure out name of dm-device from it's /dev/block/dm-XX path.
 # This is useful in case of remounting ext4 userdata into checkpointing mode,
 # since it potentially requires tearing down dm-devices (e.g. dm-bow, dm-crypto)
diff --git a/private/su.te b/private/su.te
index 690f305..7902dbf 100644
--- a/private/su.te
+++ b/private/su.te
@@ -1,6 +1,5 @@
 typeattribute su coredomain;
 
-userdebug_or_eng(`
   domain_auto_trans(shell, su_exec, su)
   # Allow dumpstate to call su on userdebug / eng builds to collect
   # additional information.
@@ -131,7 +130,6 @@ userdebug_or_eng(`
   typeattribute su hal_wifi_client;
   typeattribute su hal_wifi_hostapd_client;
   typeattribute su hal_wifi_supplicant_client;
-')
 
 # su is also permissive to permit setenforce.
 permissive su;
-- 
2.53.0

